### PR TITLE
Add Gsap Animations

### PR DESCRIPTION
**Pull request by dbertella**
This is a WIP but at least we can have a look at the diff and discuss over it

### CODE REVIEW

File: packages/client/src/utils/animations.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  const bounds = gsapElements.map((el: Element) => el.getBoundingClientRect());
+  const copy: any = {};
+  let p;
+  gsapElements.forEach((el: Element) => {
+    el._flip && el._flip.progress(1);
+    el._flip = tl;
+  });
+  changeFunc();
+  for (p in vars) {
+    p !== 'onComplete' && p !== 'delay' && (copy[p] = vars[p]);
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: This should if an `if` right?

File: packages/client/src/utils/animations.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  copy.x = (i: number, element: Element) => '+=' + (bounds[i].left - element.getBoundingClientRect().left);
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: Use backticks?

File: packages/client/src/utils/animations.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  copy.y = (i: number, element: Element) => '+=' + (bounds[i].top - element.getBoundingClientRect().top);
+  console.log(copy, bounds)
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: :raccoon:

File: packages/client/src/utils/animations.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ - ease [Ease] - controls the easing of the animation. Like "power2.inOut", or "elastic", etc.
+ - onComplete [Function] - a callback function that should be called when all the animation has completed.
+ - delay [Number] - time (in seconds) that should elapse before any of the animations begin. 
+
+This function will return a Timeline containing all the animations. 
+*/
+
+type Element = any;
+
+export function flip(elements: Element[], changeFunc: Function, vars: any = {}) {
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: Is this actually flipping or moving...?

File: packages/client/src/utils/animations.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { gsap } from 'gsap';
+
+/* 
+https://greensock.com/docs/v3/HelperFunctions
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: Do we need all that text?

File: packages/client/src/components/Players/Player.tsx [hunk 6]
```diff
@@ -10,33 +10,25 @@ import { Box, BoxProps, Grid } from 'theme-ui';
...
+export const Player = ({ player, isActive, togglePlayerActiveCard, activePlayerCard, ...rest }: Props & BoxProps) => {
   if (!player) {
     return <InvitePlayer />;
   }
   const { captured, hand } = player;
   return (
     <Box {...rest}>
       <Grid sx={{ m: 3 }} columns="1.5fr 1fr 1fr 1fr">
-        <Deck cardNumber={captured.length} scopa={player.scopa} />
+        <Deck cardNumber={captured.length} scopa={player.scopa} id="player-deck" />
```
Comment (stathismor) [inline packages/client/src/components/Players/Player.tsx:25]: These should probably live in a constants file, right?

File: packages/client/src/pages/Game.tsx [hunk 7]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
 
 const SETTEBELLO = {
   value: 7,
   suit: Suit.Golds,
 };
 
+const animateCardOnTable = (activePlayerCard: string, options: Record<string, unknown> = {}) => {
+  const cardWrap = document.querySelector(`#card_${activePlayerCard}`)?.firstChild;
+  const dropElement = document.querySelector('#drop-container');
+  console.log(cardWrap, dropElement);
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: :horse:

File: packages/client/src/pages/Game.tsx [hunk 8]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
 import { FiRotateCcw } from 'react-icons/fi';
+import { flip } from 'utils/animations';
 
 const SETTEBELLO = {
   value: 7,
   suit: Suit.Golds,
 };
 
+const animateCardOnTable = (activePlayerCard: string, options: Record<string, unknown> = {}) => {
+  const cardWrap = document.querySelector(`#card_${activePlayerCard}`)?.firstChild;
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: We should probably also move the `card_` name generation to a lib/utils file.

File: packages/client/src/pages/Game.tsx [hunk 9]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
+  console.log('Emit: Card is going to the table');
+  function moveCard() {
+    dropElement?.removeChild(dropElement.firstChild!);
+    dropElement?.appendChild(cardWrap!);
+  }
+  flip([cardWrap], moveCard, options);
+  return cardWrap;
+};
+const animateCapture = (animatedCard: ChildNode, caputeredCards: string[], options: Record<string, unknown> = {}) => {
+  const cardWraps = caputeredCards
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]:
```suggestion
  const cardWraps = capturedCards
```

File: packages/client/src/pages/Game.tsx [hunk 10]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
+    .map((cardKey) => document.querySelector(`#card_${cardKey}`)?.firstChild)
+    .concat([animatedCard]) as ChildNode[];
+  const dropElement = document.querySelector(`#player-deck`);
+  console.log(cardWraps, dropElement);
+  function moveCards() {
+    cardWraps.forEach((el) => {
+      // @ts-ignore
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: Doesn't that error go away if you assign `el` to a new const variable?

File: packages/client/src/utils/animations.ts (hunk 4 above)
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: Does this need to be initialised to `{}`. Looking below, code would not work if it's `{}`, right?

File: packages/client/src/utils/animations.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+type Element = any;
+
+export function flip(elements: Element[], changeFunc: Function, vars: any = {}) {
+  const gsapElements = gsap.utils.toArray(elements);
+  const tl = gsap.timeline({ onComplete: vars.onComplete, delay: vars.delay || 0 });
+  const bounds = gsapElements.map((el: Element) => el.getBoundingClientRect());
+  const copy: any = {};
+  let p;
+  gsapElements.forEach((el: Element) => {
+    el._flip && el._flip.progress(1);
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: What does this do...?

File: packages/client/src/utils/animations.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    el._flip = tl;
+  });
+  changeFunc();
```
Comment (stathismor) [inline packages/client/src/utils/animations.ts]: What is `changeFunc`?

File: packages/client/src/pages/Game.tsx [hunk 13]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
+  }
+  flip([cardWrap], moveCard, options);
+  return cardWrap;
+};
+const animateCapture = (animatedCard: ChildNode, caputeredCards: string[], options: Record<string, unknown> = {}) => {
+  const cardWraps = caputeredCards
+    .map((cardKey) => document.querySelector(`#card_${cardKey}`)?.firstChild)
+    .concat([animatedCard]) as ChildNode[];
+  const dropElement = document.querySelector(`#player-deck`);
+  console.log(cardWraps, dropElement);
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: :cat2:

File: packages/client/src/pages/Game.tsx [hunk 14]
```diff
@@ -14,12 +14,52 @@ import { Board } from 'components/Board';
...
+      dropElement?.appendChild(el);
+    });
+  }
+  flip(cardWraps, moveCards, {
+    ...options,
+    onComplete: () => {
+      // @ts-ignore
+      options.onComplete();
+      cardWraps.forEach((el) => {
+        dropElement?.removeChild(el);
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: What is happening here... `dropElement` is the player's deck. You first _flip_ (?), then move which appends cards to the player's deck, and then once all that is complete, you remove the cards from the deck...?

File: packages/client/src/pages/Game.tsx [hunk 15]
```diff
@@ -38,27 +78,41 @@ export const Game = ({ gameState, gameScore }: { gameState: GameState; gameScore
...
-          tableCards: activeCardsOnTable,
-        });
-        togglePlayerActiveCard(null);
-        toggleActiveCardsOnTable([]);
+        const options = {
+          onComplete: runNext,
+        };
+        const animatedCard = animateCardOnTable(activePlayerCard, options);
+        function runNext() {
+          animateCapture(animatedCard as ChildNode, activeCardsOnTable, {
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: So, do you first animate, and then emit, right...?

File: packages/client/src/pages/Game.tsx [hunk 16]
```diff
@@ -38,27 +78,41 @@ export const Game = ({ gameState, gameScore }: { gameState: GameState; gameScore
...
   const playCardOnTable = () => {
     if (activePlayerCard) {
-      console.log('Emit: Card is going to the table');
-      gameIO.emit(GameEvent.PlayerAction, roomName, {
-        action: PlayerActionType.PlayOnTable,
-        playerName: player.username,
-        card: activePlayerCard,
+      animateCardOnTable(activePlayerCard, {
+        onComplete: () => {
+          console.log('tx complete', activePlayerCard);
```
Comment (stathismor) [inline packages/client/src/pages/Game.tsx]: :mosquito:

File: packages/client/src/utils/animations.ts (hunk 5 above)
Comment (dbertella) [reply packages/client/src/utils/animations.ts]: I think it's good to keep it as a reference, it comes from this function is built better to keep it than to have to find it all the time right?

File: packages/client/src/components/Players/Player.tsx (hunk 6 above)
Comment (dbertella) [reply packages/client/src/components/Players/Player.tsx:25]: yes I was thinking to do that too, but for now it's just to show off

File: packages/client/src/utils/animations.ts (hunk 11 above)
Comment (dbertella) [reply packages/client/src/utils/animations.ts]: 🤷‍♂️

File: packages/client/src/pages/Game.tsx (hunk 15 above)
Comment (dbertella) [reply packages/client/src/pages/Game.tsx]: yes emit happens in the onConplete cb

**Pull request merged**